hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java

```diff
@@ -821,6 +865,38 @@ protected void finalizeWrite(HoodieTable<T, I, K, O> table, String instantTime,
     }
   }
 
+  private HoodieCommitMetadata resolveWriteConflictIfAny(HoodieTable<T, I, K, O> table, final String instantTime,
+                                            HoodieCommitMetadata thisCommitMetadata) {
+    Long currentTimeInSecs = System.currentTimeMillis() / 1000;
+    ConflictResolutionStrategy resolutionStrategy = config.getWriteConflictResolutionStrategy();
+    Option<HoodieInstant> lastCompletedInstantBeforeWriterStarted = this.latestWriteInstantCompletedBeforeWriter.get();
+    String lastInstantTimestamp = lastCompletedInstantBeforeWriterStarted.isPresent()
+        ? lastCompletedInstantBeforeWriterStarted.get().getTimestamp() : "0";
+    Stream<HoodieInstant> instantStream = table.getActiveTimeline()
+        .getAllCommitsTimeline()
+        .getCommitsAndCompactionTimeline()
+        .filterCompletedInstants()
+        .findInstantsInRange(lastInstantTimestamp, String.valueOf(currentTimeInSecs))
+        .getInstants();
+
+    LOG.info("Current eligible instants during write oHoodieWriteConfigf instant " + instantTime + " = "
+        + instantStream.collect(Collectors.toList()));
+
+    boolean hasConflict = instantStream.anyMatch(instant -> {
+      try {
+        return resolutionStrategy.hasConflict(thisCommitMetadata, HoodieCommitMetadata.fromBytes(
+            table.getActiveTimeline().getInstantDetails(instant).get(), HoodieCommitMetadata.class));
+      } catch (IOException io) {
+        throw new HoodieCommitException("Unable to determine if conflict exists", io);
+      }
+    });
+    HoodieCommitMetadata newMetadataAfterConflictResolution = thisCommitMetadata;
+    if (hasConflict) {
+      newMetadataAfterConflictResolution = resolutionStrategy.resolveConflict(config, hadoopConf);
```

Comment:
we should probably have an API to resolveConflict(instant1, instant2) ? i.e like our preCombine method? that will give resolutionStrategy implementors to just reason between two instants and whether they conflict or not, and if they do, then munge the commit metadata somehow.

Reply:
Done